### PR TITLE
Some enhancements of conversion script

### PR DESCRIPTION
**Pull request by sampsongao**
- Update node-addon-api module name
- Turn on c++ exception in binding.gyp
- Reorder some conversions so that more specific changes come first

### CODE REVIEW

**Comment by mathiask88** [comment]
I tried it with my [addon](https://github.com/mathiask88/node-snap7) but the result is just messy and doing it manually would be much easier I think. Things are replaced multiple times, regex strings are inserted and even some code is getting lost. I know this kind of conversion scripts will never do the job and are just some helper, but I don't understand my own code after using it :D

**Comment by sampsongao** [comment]
I just commit a new change to it to get rid of the `Napi::Napi::` repetition. However, for now we are not taking care of indentation. That could be why it looks messy. There are some Nan call which is not easily convertible to Napi, so there always some manual work left for now. For example, 
```
NAN_MODULE_INIT(S7Client::Init) {

  Nan::HandleScope scope;
 
  v8::Local<v8::FunctionTemplate> tpl;
  tpl = Nan::New<v8::FunctionTemplate>(S7Client::New);

  v8::Local<v8::String> name = Nan::New<v8::String>("S7Client")
    .ToLocalChecked();
  
  tpl->SetClassName(name);
  tpl->InstanceTemplate()->SetInternalFieldCount(1);
 
  // Setup the prototype
  // Control functions
  Nan::SetPrototypeMethod(
      tpl
    , "Connect"
    , S7Client::Connect);
  Nan::SetPrototypeMethod(
      tpl
    , "ConnectTo"
    , S7Client::ConnectTo);
  ...
}

...

S7Client::S7Client() {
  snap7Client = new TS7Client();
  uv_mutex_init(&mutex);
}

...

NAN_METHOD(S7Client::New) {
  ...
}
```

Should convert to 
```
void S7Client::Init(Napi::Env env, Napi::Object exports, Napi::Object module) {
  Napi::HandleScope scope(env);

  Napi::Function tpl = DefineClass(env, "S7Client",
    InstanceMethod("Connect", S7Client::Connect),
    InstanceMethod("ConnectTo", S7Client::ConnectTo),
    ...
  );

...

S7Client::new() {
  snap7Client = new TS7Client();
  uv_mutex_init(&mutex);
}

...

S7Client::S7Client(const Napi::CallbackInfo & info) : Napi::ObjectWrap (info) {
  new();
  ...
}

```

**Comment by mhdawson** [comment]
@mathiask88 if there were 1 or 2 things that would improve the conversion for you what would they be ?

**Comment by mathiask88** [comment]
@mhdawson I think a simple conversion script like this and a good migration doc is enough (It was for NAN 1->2). There is no need for a super complex conversion tool, but the last version of the script messed up the code for me. The updated script is much better than the old one, but I have some more things that should be possible to fix.

`info[1]->Int32Value()`
becomes
`info[4].Int32Value()`
but should be
`info[4].As<Number>().Int32Value()`

`info.GetReturnValue().SetUndefined();`, `info.Holder()` are not handled yet.

And would it make sense to add a `auto env = info.Env();` in the next line after a `NAN_METHOD(..)`?

`[ /Nan::(Undefined|Null|True|False)\(\)/g, 'env.$1()' ],` `True` and `False` are no members from `Napi::Env`.

**Comment by sampsongao** [comment]
We might want to add env.True() and env.False(). @jasongin

File: tools/conversion.js [hunk 1]
```diff
@@ -89,22 +90,28 @@ var SourceFileOperations = [
   [ /->IsInt32\(\)/g, '.IsNumber()' ],
 
 
+  [ /(.+?)->BooleanValue()/g, '$1.As<Napi::Boolean>().BooleanValue()' ],
+  [ /(.+?)->Int32Value()/g, '$1.As<Napi::Number>().Int32Value()' ],
+  [ /(.+?)->Uint32Value()/g, '$1.As<Napi::Number>().Uint32Value()' ],
+  [ /(.+?)->IntegerValue()/g, '$1.As<Napi::Number>().Int64Value()' ],
+  [ /(.+?)->NumberValue()/g, '$1.As<Napi::Number>().DoubleValue()' ],
+
```
Comment (mathiask88) [inline tools/conversion.js]: Missing escapes here `\(\)`

File: tools/conversion.js [hunk 2]
```diff
@@ -10,182 +10,211 @@ if (!dir) {
...
+  // ex. Nan::To<int64_t>(info[0]) to info[0].As<Napi::Number>().Int64Value()
+  [ /Nan::To<int64_t>\((.+?)\)/g, '$1.As<Napi::Number>().Int64Value()' ],
+  // ex. Nan::To<float>(info[0]) to info[0].As<Napi::Number>().FloatValue()
+  [ /Nan::To<float>\((.+?)\)/g, '$1.As<Napi::Number>().FloatValue()' ],
+  // ex. Nan::To<double>(info[0]) to info[0].As<Napi::Number>().DoubleValue()
+  [ /Nan::To<double>\((.+?)\)/g, '$1.As<Napi::Number>().DoubleValue()' ],
+
+  [ /Nan::New\((\w+)\)->HasInstance\((\w+)\)/g, '$2.InstanceOf($1.Value())' ],
+
+  [ /Nan::Get\(([^)]+),\s*/gm, '($1).Get(' ],
```
Comment (mathiask88) [inline tools/conversion.js]: I think `/Nan::Get\(([^,\s]+)\s*,\s*/gm` is better for nested stuff. And maybe add the same thing for `Nan::Has()`

File: tools/conversion.js [hunk 3]
```diff
@@ -10,182 +10,211 @@ if (!dir) {
...
+  [ /\.Get\([\s|\\]*Nan::New\(([^)]+)\)\)/gm, '.Get($1)' ],
+
+  [ /Nan::Set\(([^,]+),/gm, '($1).Set(' ],
```
Comment (mathiask88) [inline tools/conversion.js]: same regex as `Get()`

File: tools/conversion.js [hunk 4]
```diff
@@ -10,182 +10,211 @@ if (!dir) {
...
-  [/Nan::New<Integer>\((.+?)\)/g, 'Napi::Number::New(env, $1)'],
-  [/Nan::New\(([0-9\.]+)\)/g, 'Napi::Number::New(env, $1)'],
-  [/Nan::New\("(.+?)"\)/g, 'Napi::String::New(env, "$1")'],
-  [/Nan::New<(.+?)>\(\)/g, 'Napi::$1::New(env)'],
-  [/Nan::New<(.+?)>\(/g, 'Napi::$1::New(env, '],
-  [/Nan::NewBuffer\(/g, 'Napi::Buffer<char>::New(env, '],
+  [ /Nan::New<v8::Integer>\((.+?)\)/g, 'Napi::Number::New(env, $1)' ],
+  [ /Nan::New<Integer>\((.+?)\)/g, 'Napi::Number::New(env, $1)' ],
+  [ /Nan::New\(([0-9\.]+)\)/g, 'Napi::Number::New(env, $1)' ],
+  [ /Nan::New<v8::String>\("(.+?)"\)/g, 'Napi::String::New(env, $1)' ],
```
Comment (mathiask88) [inline tools/conversion.js]: Missing quotes `"$1"` here?!

**Comment by mhdawson** [comment]
@mathiask88 will land this version and we'll see what we can do about your other suggestions @sampsongao can you open a new issue to track covering off the suggestions in this comment: https://github.com/nodejs/node-addon-api/pull/77#issuecomment-314996453

**Comment by mhdawson** [comment]
Landed as 27e99520caf005c2d82242651573d31fc1675d81

**Pull request closed without merging**